waspc/src/Wasp/ExternalConfig/PackageJson.hs

```diff
@@ -0,0 +1,127 @@
+
+type PackageName = String
+
+type PackageVersion = String
+
+instance FromJSON PackageJson
+
+analyzePackageJsonContent :: Path' Abs (Dir WaspProjectDir) -> IO (Either [CompileError] PackageJson)
+analyzePackageJsonContent waspProjectDir = runExceptT $ do
+  packageJsonFile <- ExceptT findPackageJsonFileOrError
+  packageJson <- ExceptT $ readPackageJsonFile packageJsonFile
+  ExceptT $ validatePackageJson packageJson
+  where
+    findPackageJsonFileOrError = maybeToEither [fileNotFoundMessage] <$> findPackageJsonFile waspProjectDir
+    fileNotFoundMessage = "Couldn't find the package.json file in the " ++ toFilePath waspProjectDir ++ " directory"
+
+validatePackageJson :: PackageJson -> IO (Either [CompileError] PackageJson)
+validatePackageJson packageJson =
+  return $
+    if null packageJsonErrors
+      then Right packageJson
+      else Left packageJsonErrors
+  where
+    packageJsonErrors =
+      concat
+        [ -- Wasp needs the Wasp SDK to be installed in the project.
+          validate ("wasp", "file:.wasp/out/sdk/wasp", RequiredPackage),
+          -- Wrong version of Prisma will break the generated code.
+          validate ("prisma", show prismaVersion, RequiredDevPackage),
+          -- Installing the wrong version of "react-router-dom" can make users believe that they
+          -- can use features that are not available in the version that Wasp supports.
+          validate ("react-router-dom", show reactRouterVersion, OptionalPackage)
+        ]
+    validate = validatePackageInDeps packageJson
+
+findPackageJsonFile :: Path' Abs (Dir WaspProjectDir) -> IO (Maybe (Path' Abs File'))
+findPackageJsonFile waspProjectDir = findFileInWaspProjectDir waspProjectDir packageJsonInWaspProjectDir
+
+readPackageJsonFile :: Path' Abs File' -> IO (Either [CompileError] PackageJson)
+readPackageJsonFile packageJsonFile = do
+  byteString <- IOUtil.readFileBytes packageJsonFile
+  return $ maybeToEither ["Error parsing the package.json file"] $ Aeson.decode byteString
+
+data PackageValidationType = RequiredPackage | RequiredDevPackage | OptionalPackage
+
+validatePackageInDeps :: PackageJson -> (PackageName, PackageVersion, PackageValidationType) -> [CompileError]
+validatePackageInDeps packageJson (packageName, expectedPackageVersion, validationType) =
+  case map (M.lookup packageName) depsToCheck of
+    (Just actualPackageVersion : _) ->
+      if actualPackageVersion == expectedPackageVersion
+        then []
+        else [incorrectVersionMessage]
+    _rest -> case validationType of
+      RequiredPackage -> [requiredPackageMessage "dependencies"]
+      RequiredDevPackage -> [requiredPackageMessage "devDependencies"]
+      OptionalPackage -> []
+  where
+    depsToCheck = case validationType of
+      RequiredPackage -> [dependencies packageJson]
+      RequiredDevPackage -> [devDependencies packageJson]
+      -- Users can install packages that don't need to be strictly in dependencies or devDependencies
```

Comment:
We don't control where the user installs the dep: production or development deps. If they install some deps like `react-router-dom` either as production or development - they will mess up their project somewhat. Your name is better `HasExactVersionIfListed` -> if the user added it - make sure it has the correct version.

Reply:
Right, but is it possible that we care about that dep only if they install it as non-dev, or maybe dev dep? Maybe we need react-router-dom to be specific version if they install it as normal dep, but we don't care if they install it as dev-dep?

Following up on this: we require them to have `react` installed as non-dev dep with specific version. What if they also install it as dev dep with some other version? Is that something we want to throw warning upon?

Yeah this is the direction I am inquirying about: seems to me like sometimes we care if it is dev or non-dev dep, but sometimes we say "any of those, dev or non-dev", and it doesn't seem super clear to me how we are making those decisions, seems like we haven't thought it through completely. Although I get why, it is a bit confusing / hard to define, but it is probably worth giving it a try and thinking it all the way through.

Reply:
> Right, but is it possible that we care about that dep only if they install it as non-dev, or maybe dev dep? Maybe we need react-router-dom to be specific version if they install it as normal dep, but we don't care if they install it as dev-dep?

`npm install` doesn't care if it's a dev dep or prod dep. `npm ci` cares, but users don't use that. So for our use case "let's prevent users from messing up their Wasp project" - we need to validate both equally. 

> Following up on this: we require them to have react installed as non-dev dep with specific version. What if they also install it as dev dep with some other version? Is that something we want to throw warning upon?

That's not possible. If you try installing React as a dev dep, it will just move from deps to dev deps.

To conclude: some deps have strict location (dep vs. dev dep) and version requirements but for some the location is irrelevant due to `npm install` not caring where it's installed (it all goes to one `node_modules`) and only the version needs to be validated.

Reply:
Oooooh ok, I actually had this false idea that devDependencies are somehow better isolated from the normal ones, I get now how it works. Thanks, ok this makes sense then! Feel free to resolve this comment.